### DISCUSSION TITLE
Is `opentelemetry_tracing_layer` needed when there is `TraceLayer::new_for_http()`?

### DISCUSSION THREAD

**Discussion by lcmgh** [tokio-rs/axum#1733] [Q&A] [answered]
I have setup my app in a way that it can be instrumented with `tracing` crates `#[instrument]` macro.

```rust
let tracer = init_tracer().unwrap();

tracing_subscriber::registry()
    .with(tracing_subscriber::EnvFilter::new(
        std::env::var("RUST_LOG").unwrap_or_else(|_| "myapp=info,tower_http=info".into()),
    ))
    .with(tracing_subscriber::fmt::layer())
    .with(tracing_opentelemetry::layer().with_tracer(tracer))
    .init();
```

My router
```rust
..
.layer(opentelemetry_tracing_layer())
.layer(init_session_layer())
.layer(TraceLayer::new_for_http());
```

However the web api calls do not end up in my tracing suite. I am also lacking of the general understanding if there is a need for the `opentelemetry_tracing_layer` crate when there is `TraceLayer::new_for_http()`?

thx for any clarification

**Comment by davidpdrsn** [comment]
axum doesn't have any `opentelemetry_tracing_layer`. Does that come from some external crate or something? If so, it depends on what it does I guess.

**Comment by lcmgh** [comment, marked as answer]
`opentelemetry_tracing_layer` (`axum_tracing_opentelemetry::middleware::trace_extractor
`) comes from the `axum-tracing-opentelemetry` crate. It internally adds the `TraceLayer::new_for_http()`. 

I am just struggling to get it working, see https://github.com/davidB/axum-tracing-opentelemetry/issues/44.

**Reply by davidpdrsn** [reply]
I'm not familiar with that library but good that you opened an issue there as well. I'm afraid I can't offer much help.

**Reply by lcmgh** [reply]
Solved it already. I think this crate is quite important to enable Axum usage in corporate environments with otel tracing requirements.
